jack module: init

gnidorah opened the pull request:
###### Motivation for this change

Currently tested applications with default module settings:
1. Bitwig, Ardour, MuseScore, ADLplug standalone via native JACK
2. VLC, Firefox, Chrome, QMMP, aplay, uade123, openmpt123, munt via pcm plugin and loopback device
3. hardware.pulseaudio.enable both true and false

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

gnidorah commented:
@oxij The only sad thing is that I have not found a way to make jack1 work correctly in promiscuous mode:
http://lists.jackaudio.org/pipermail/jack-devel-jackaudio.org/2019-March/001876.html
https://github.com/jackaudio/jack1/issues/87

gnidorah commented:
cc @bobvanderlinden for alsa-plugins change

oxij commented:
Nice! Thanks for putting effort into NixOSifying this.

The first patch LGTM.

The second patch:

- first line of the commit message should be "nixos/jack: init" or similar,
- `extraPackages` option looks superfluous to me,
- `enable32BitAlsaPlugins` should be an option, like `hardware.pulseaudio.support32Bit` and `hardware.opengl.driSupport32Bit`, the current definition is always `true` AFAICS.
- "/home/jackaudio" -> "/var/lib/jackaudio",
- that user needs a static uid, I think,
- you can't use `createHome` there, the home directory needs to be created on every startup if it does not exist, else things are very easy to break.

From the top of my head, see the tor service module for how to do all these user- and home-related things properly.

I would also rename

- `services.jack.pcmPlugin` -> `services.jack.alsa.enable`
- `services.jack.dbusSupport` -> services.jack.dbus.enable`
- both `startupOptions` -> `extraOptions`

as common in other modules.

I would also add `services.jack.a2jmidid.package` option for completeness and put all `jackd` options under `services.jack.jackd` for homogeneity.

> The only sad thing is that I have not found a way to make jack1 work correctly in promiscuous mode:

Hm, I had it working before, I'll play with it sometime later.

gnidorah commented:
@oxij 

Thank you! Applied changes.

> you can't use `createHome` there, the home directory needs to be created on every startup if it does not exist, else things are very easy to break

Tor module uses createHome too.

> I would also add `services.jack.a2jmidid.package` option for completeness

We only have one version and also JACK1 has this functional builtin

> Hm, I had it working before, I'll play with it sometime later.

Would be really glad if you let me know how to make it work (no necessity to test this PR, just an instruction on how to kick it will be enough). For me only control functions do work with jack1 in promiscuous mode. If I try to user programs there are errors "subgraph starting at ADLplug timed out (subgraph_wait_fd=9, status = 0, state = Triggered, pollret = 0 revents = 0x0)". If I run jack client from the same user as jack server user then everything works.

Yarny0 commented:
> - you can't use `createHome` there, the home directory needs to be created on every startup if it does not exist, else things are very easy to break.

Just a hint (hope that helps): The systemd directives `StateDirectory` and `StateDirectoryMode` can be used to create a "home" directory for a service in `/var/lib` (or in the user's home directory if it is a user service). The directory will be owned by the user account that is used for the service and it persists when the service or the machine is restarted. See systemd.exec(5) for details. If the service actually needs a $HOME environment variable, it must be defined with the Environment systemd directive.

gnidorah commented:
@Yarny0 applied, thanks!

gnidorah commented:
Will reopen after some rework

gnidorah commented:
@oxji Hi! Did you get a chance to look into jack1 promiscuous mode?
Meanwhile I've reworked this module to make it more robust and simpler to maintain.
1. Removed user service. Because of #21460 and also for simplicity and maintainability
2. Added loopback device option in addition to ALSA PCM plugin. Makes setup work in sandboxed environments like Steam
3. Added session option so user could do initialization here. Default session just starts a2jmidid if JACK2 is used
4. Used jack_wait and jack_lsp to check for server/port availability instead of unreliable delays in services
5. JACK service is now brought not by sound.target but by udev rule when real sound card appears in order not to trigger on loopback device availability

Infinisil commented on nixos/modules/services/audio/jack.nix:
```diff
@@ -0,0 +1,275 @@
+    services.jack = {
+      jackd = {
+        enable = mkEnableOption ''
+          JACK Audio Connection Kit. You need to add yourself to "jackaudio" group
```
to *the* "jackaudio" group

Infinisil commented on nixos/modules/services/audio/jack.nix:
```diff
@@ -0,0 +1,275 @@
+        };
+
+        session = mkOption {
+          type = types.string;
```
Should be `types.lines`

Infinisil commented on nixos/modules/services/audio/jack.nix:
```diff
@@ -0,0 +1,275 @@
+        session = mkOption {
+          type = types.string;
+          default = ''
+            ${lib.optionalString bridgeNeeded "${pkgs.a2jmidid}/bin/a2jmidid -e &"}
```
Set this in the config section like this:
```nix
{
  config = {
    services.jack.jackd.session = ''
      ${lib.optionalString ...}
    '';
  };
}
```

Otherwise this default will get cleared if users set a value themselves

Infinisil commented on nixos/modules/services/audio/jack.nix:
```diff
@@ -0,0 +1,275 @@
+          default = false;
+          description = ''
+            Whether to include the 32-bit alsa plugins libraries in the system or not.
+            This is only useful on 64-bit systems and currently limited to x86_64-linux.
```
What's the use case of this?

gnidorah replied:
@Infinisil 32bit applications using alsa (instead of jack) will be able to play sound on 64bit system when this is enabled.

Infinisil replied:
Maybe you could add that to the description. This wasn't obvious to me.

Infinisil commented on nixos/modules/services/audio/jack.nix:
```diff
@@ -0,0 +1,275 @@
+        };
+
+        config = mkOption {
+          type = types.string;
```
Should be `types.lines`

Infinisil commented on nixos/modules/services/audio/jack.nix:
```diff
@@ -0,0 +1,275 @@
+          '';
+          # https://alsa.opensrc.org/Jack_and_Loopback_device_as_Alsa-to-Jack_bridge#id06
+          default = ''
+            pcm.loophw00 {
```
It also might make sense to put this in the config section so users can change it easily without losing all these defaults. The disadvantage of putting it in the config section is that it won't show up in the manual.

Infinisil commented on nixos/modules/services/audio/jack.nix:
```diff
@@ -0,0 +1,275 @@
+        session = mkOption {
+          type = types.string;
+          default = ''
+            alsa_in -j cloop -dcloop &
```
Same thing, should be `types.lines` and be in the config section to not lose these defaults

Infinisil commented on nixos/modules/services/audio/jack.nix:
```diff
@@ -0,0 +1,275 @@
+     assertions = [
+        {
+          assertion = !(cfg.alsa.enable && cfg.loopback.enable);
+          message = "Both alsa and loopback options shouldn't be used at the same time.";
```
You need to add more context, mention that this assertion is about JACK

Infinisil commented on nixos/modules/services/audio/jack.nix:
```diff
@@ -0,0 +1,275 @@
+        description = "JACK Audio Connection Kit";
+        serviceConfig = {
+          User = "jackaudio";
+          ExecStart = "${cfg.jackd.package}/bin/jackd ${concatStringsSep " " cfg.jackd.extraOptions}";
```
Use `lib.escapeShellArgs` instead of `concatStringsSep " "`

gnidorah replied:
@Infinisil Hmm, this way it generates `ExecStart=/nix/store/inqg0qsbd5z0i8l4515n3h800dm6di82-jack2-1.9.12/bin/jackd '-dalsa' '--device hw:1'` instead of `ExecStart=/nix/store/inqg0qsbd5z0i8l4515n3h800dm6di82-jack2-1.9.12/bin/jackd -dalsa --device hw:1` and so jackd fails to start

Infinisil replied:
Hmm I see, you'd have to change the declaration to `[ "-dalsa" "--device" "hw:1" ]`. Often CLI's support leaving out the space even, so maybe you can do `[ "-dalsa" "--devicehw:1" ]`.

gnidorah commented:
@Infinisil Thank you for review! Could you please tell me more about moving defaults to config?
What if user wants to drop these defaults? Should he use `lib.mkForce` then?
I personally keep defaults with `options` like this
```nix
  services.xserver.windowManager.i3.extraPackages = with pkgs;
    options.services.xserver.windowManager.i3.extraPackages.default ++ [ spectacle ];

  programs.sway.extraPackages = with pkgs;
    options.programs.sway.extraPackages.default ++ [ qt5.qtwayland grim ];
```
Just can't conclude which approach is better:
- to set defaults in config, so user likely needs `mkForce` if he wants to overwrite them
- or to set defaults in options and let user use `options...default` is he wants to extend defaults

Infinisil commented:
I think if the default makes sense to always be present, because otherwise the module doesn't properly work, then setting it in the config section and requiring the user to use mkForce if they really want to remove that is good.

On the other hand for defaults that aren't necessary for the module to work, and users are encouraged to change them, then assigning a default in the option makes more sense. Note that the users could still keep those defaults by assigning their values with `mkOptionDefault`, which is a bit weird, but simpler than using the `options` argument, but sometimes also doesn't work.

(The option value gets determined with a priority system, the assignments of highest priority get merged together to result in the final value, mkForce has high priority, mkOptionDefault very low priority, mkDefault low priority, and without any of those it gets a default priority in the middle).

I've written down something similar for NixOS/rfcs#42: https://github.com/Infinisil/rfcs/blob/config-option/rfcs/0042-config-option.md#defaults

In this case I'd think putting the defaults in the config section always makes more sense, because these default values encode part of the functionality of this module, and without them certain options couldn't do their thing.

The biggest annoyance is that these defaults don't show up in the manual. In the future this might get fixed.

gnidorah commented:
@Infinisil Got it, thank you! I've applied requested changes.

Infinisil commented:
Hm I just wanted to test this, but I'm not sure it works as intended. I enabled `services.jack.jackd.enable = true;`, but because there's no `wantedBy = [ "multi-user.target" ]` or so for the jackd service, it doesn't even start. Is that right? How do I test this?

gnidorah commented:
@Infinisil it will start only on reboot by udev rule, when real sound card will be added, sound.target is not suitable, since that way loopback device could be initialized before real sound card, and so service will fail. udev rule doesn't trigger on loopback device

Commit:
nixos/jack: init

gnidorah commented:
Could someone finally merge this?

Infinisil commented:
While I'd have liked for another person to test it, especially somebody who wants to use jack, you seem to have tested it well enough already, so let's merge.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/firewire-audio-not-working/3976/4
